### PR TITLE
feature: Add inlay hints for named parameters

### PR DESCRIPTION
**Pull request by tgodzik**
Scala 2 version of https://github.com/scalameta/metals-feature-requests/issues/184

### CODE REVIEW

**Review by kasiaMarek — COMMENTED** [review]
I feel like it would be cool to have inlay hints for only "useful" named params. Not necessarily instead of this but maybe additionally. We could avoid showing named params:
1. for single letter param names (or `xs`, `ys`), since those are usually meaningless
2. when param name and passed argument have the same name

File: tests/cross/src/test/scala/tests/pc/InlayHintsSuite.scala [hunk 1]
```diff
@@ -336,7 +323,7 @@ class InlayHintsSuite extends BaseInlayHintsSuite {
        |}
        |""".stripMargin,
     """|object Main {
-       |  val x/*: (Int<<scala/Int#>>, Int<<scala/Int#>>)*/ = Tuple2/*[Int<<scala/Int#>>, Int<<scala/Int#>>]*/(1, 2)
+       |  val x/*: (Int<<scala/Int#>>, Int<<scala/Int#>>)*/ = Tuple2/*[Int<<scala/Int#>>, Int<<scala/Int#>>]*/(/*_1 = */1, /*_2 = */2)
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply tests/cross/src/test/scala/tests/pc/InlayHintsSuite.scala:326]: True, but I want to keep it consistent. It's also rare that Tuple is used explicitly.

**Comment by tgodzik** [comment]
> I feel like it would be cool to have inlay hints for only "useful" named params. Not necessarily instead of this but maybe additionally. We could avoid showing named params:
> 
> 1. for single letter param names (or `xs`, `ys`), since those are usually meaningless

This really depends on a use case. Most of the times I guess yeah, but sometimes it might be useful for some symbolic names.

> 2. when param name and passed argument have the same name

I want to keep it consistent, if we don't show named params in some cases, users might actually think they are broken.

File: mtags/src/main/scala-2/scala/meta/internal/pc/PcInlayHintsProvider.scala [hunk 2]
```diff
@@ -142,6 +150,60 @@ final class PcInlayHintsProvider(
...
+              if isRealApply(fun) &&
+                /* We don't want to show named parameters for unapplies*/
+                args.exists(arg => arg.pos.isRange && arg.pos != fun.pos) &&
+                /* It's most likely a block argument, even if it's not it's
+                   doubtful that anyone wants to see the named parameters */
+                !isSingleBlock(args) =>
+            val params = fun.tpe.params
+            Some(args.zip(params).collect {
+              case (arg, param)
+                  if !fun.symbol.isJavaDefined && // We don't have a quick way of getting java param names
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply mtags/src/main/scala-2/scala/meta/internal/pc/PcInlayHintsProvider.scala]: changed!

File: mtags/src/main/scala-2/scala/meta/internal/pc/PcInlayHintsProvider.scala [hunk 3]
```diff
@@ -142,6 +150,60 @@ final class PcInlayHintsProvider(
...
+                    !namedArgCache.exists(p => p._1.includes(arg.pos)) =>
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply mtags/src/main/scala-2/scala/meta/internal/pc/PcInlayHintsProvider.scala]: I hoped we would have a utility method, but there is nothing after typer to see if it was named argument. The only case this works is when we have all the arguments named or different order.

I switched to simple checking if '=' is last non whitespace char. That should simple enough and work in most cases.

File: tests/cross/src/test/scala/tests/pc/InlayHintsSuite.scala [hunk 4]
```diff
@@ -1075,6 +958,21 @@ class InlayHintsSuite extends BaseInlayHintsSuite {
...
+  check(
+    "by-name-regular-named",
+    """|object Main{
+       |  def foo(x: => Int, y: Int, z: => Int)(w: Int, v: => Int): Unit = ()
+       |  foo(1, 2, 3)(4, 5)
+       |}
+       |""".stripMargin,
+    """|object Main{
+       |  def foo(x: => Int, y: Int, z: => Int)(w: Int, v: => Int): Unit = ()
+       |  foo(/*x = => */1, /*y = */2, /*z = => */3)(/*w = */4, /*v = => */5)
```
Comment (tgodzik) [inline tests/cross/src/test/scala/tests/pc/InlayHintsSuite.scala:974]: @harpocrates I noticed that our PR interact with each other, what do you think about the mix of named parameters and by name ones? Alternatively, we could drop '=' and have:
```suggestion
       |  foo(/*x => */1, /*y = */2, /*z => */3)(/*w = */4, /*v => */5)
```

Comment (harpocrates) [reply tests/cross/src/test/scala/tests/pc/InlayHintsSuite.scala:974]: I think seeing `= =>` is probably the most reasonable thing - unfortunately in scala `x => 1` starts looking like a lambda w/ arg `x`. I don't have a strong opinion though...

**Comment by harpocrates** [comment]
On dropping some parameter inlay-hints when they'd probably be noise: I don't really have an opinion to add, but wanted to mention this is the approach `rust-analyzer` took. Here's the function they used to bake in all of their heuristics:

https://github.com/rust-lang/rust-analyzer/blob/16745db3f763b6a1a315a91e8f4191b6a7a9da63/crates/ide/src/inlay_hints/param_name.rs#L111

**Comment by tgodzik** [comment]
> On dropping some parameter inlay-hints when they'd probably be noise: I don't really have an opinion to add, but wanted to mention this is the approach `rust-analyzer` took. Here's the function they used to bake in all of their heuristics:
> 
> https://github.com/rust-lang/rust-analyzer/blob/16745db3f763b6a1a315a91e8f4191b6a7a9da63/crates/ide/src/inlay_hints/param_name.rs#L111

Thanks! Though I think we can add the heuristics later when testing the feature and user have some specific ideas. For now I will leave it as straightforward as possible.

File: tests/cross/src/test/scala/tests/pc/InlayHintsSuite.scala [hunk 5]
```diff
@@ -1116,6 +1014,217 @@ class InlayHintsSuite extends BaseInlayHintsSuite {
...
+       |  }
+       |
+       |  def downloadSemanticdbJavac2: List[Path] = {
+       |    downloadDependency(
+       |      Dependency.of(
+       |        /*organization = */"com.sourcegraph",
+       |        /*name = */"semanticdb-javac",
+       |        /*version = */"BuildInfo.javaSemanticdbVersion",
+       |      ),
+       |      None,
```
Comment (tgodzik) [inline tests/cross/src/test/scala/tests/pc/InlayHintsSuite.scala:1221]: This becomes much more complex now since all the parameters are now synthetic if we reorder the params in any way (though the flag is not set in reality). I might work on it later, but I will just do a follow up issue before mergeing. TODO

File: mtags/src/main/scala-2/scala/meta/internal/pc/PcInlayHintsProvider.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (kasiaMarek) [reply mtags/src/main/scala-2/scala/meta/internal/pc/PcInlayHintsProvider.scala]: Oh, if that information is lost in typed tree, the `namedArgCache` approach was fine.

[the remark this answers is not in the document]
Comment (tgodzik) [reply mtags/src/main/scala-2/scala/meta/internal/pc/PcInlayHintsProvider.scala]: I think this currently works ok, I added tests around some edge case. If there is a bug it will not be a huge problem

File: mtags/src/main/scala-2/scala/meta/internal/pc/PcInlayHintsProvider.scala [hunk 6]
```diff
@@ -150,6 +159,71 @@ final class PcInlayHintsProvider(
...
+
+    private def isDefaultArgument(arg: Tree) = {
+      arg.symbol != null && arg.symbol.isDefaultGetter
+    }
+    private def isNamedArg(arg: Tree) = {
+      def loop(i: Int): Boolean = {
+        if (text(i) == '=') true
+        else if (
+          text(i) == ' ' || text(i) == '\t'
+          || text(i) == '\n' || text(i) == '\r'
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply mtags/src/main/scala-2/scala/meta/internal/pc/PcInlayHintsProvider.scala]: Good catch!
